### PR TITLE
Chapter 6 production polish

### LINKED ISSUE #9: classification2
Branch: [classification2](https://github.com/UBC-DSCI/introduction-to-datascience-python/tree/classification2)
PR opened: PR #10 , the [deployed Chapter](https://gloriawyy.github.io/classification2/classification2.html)

### LINKED ISSUE #70: Suggestions for Ch6 - Classification 2
- [x] np.random.seed is not recommended anymore
    - [x] Instead we shuold create a random state object and use it methods
    - [x] However, I don't think we will ever need to use np.random.seed. With both sklearn methods and pandas `sample` we will always control randomness via the `random_state` parameter.
    - [x] I think it is more confusing than helpful to also talk about seeds here, and we can just show the effect of a different `random_state` when doing the `train_test` split instead.
- [x] "only function that uses randomness in **Python**"
- [x] We haven't taught how to use a lambda function like this. We could since it is useful and very flexible, or if we think it is too complex, we could teach how to perform the same operation either via `cancer["Class"].map({'M': 'Malignant', 'B': 'Benign'})` (which would be convenient but a new thing) or simply filtering: `cancer.loc[cancer['Class'] == 'M', 'Class'] = 'Malignant'` + the benign version (which is less elegant but builds on what they have learned already)
- [x] We are setting the seed in this cell but there is no randomness to control...
- [x] We don't need to set the color range
- [x] Downgrade to pandas <1.5 to avoid warning until next Altair release ` FutureWarning: iteritems is deprecated and will be removed in a future version. Use .items instead.`
- [ ] Explain why it is important to use `stratify` and `shuffle`, not just mention what they do (include a short example for each in a sentence or so)
- [x] This sentence is meant to include the actual number of observations "We can see from .info() in the code above that the training set contains observations, while the test set contains observations"
- [x] The `.info` method always prints columns vertically, not just when there is a large number of them.
- [x] We should not use `groupby` + `count` here. That creates a count for each column, so it doesn't generalize well. Instead we should use `.groupby` + `size` OR just use `value_counts(normalize=True)` which means we don't have to divide by the length after (btw we should get dataframe length via `df.shape[0]` not using `len` (applies to the entire book)). I don't think there is a `normalize` option to `size`, but can't double check because I'm in the air
    - [x] We also shouldn't create a new dataframe unless we are using it again later.
- [x] This sentence is meant to include actual numbers: "we see about % of the training data are benign and % are malignant"
- [x] Imports seems to be missing for all sklearn stuff.
- [x] Could we just call it "model" instead of "model specification" everywhere? Make it flow better so it is easier to read.
- [x] Some introduction to why we use captical `X` and lower case `y`?
    - [ ] `X` should be created as `cancer_train.drop(columns='Class')`, which is more generalizable
    - [ ] We could possibly even skip this throughout the book and just pass the dataframe?
        - [ ] If we keep it, we should call it `X_train` and do the same for the test data.
        - [ ] I think we should probably keep the assignment because we used them a few times later, like when creating `X_tune` / `y_tune` in grdisearch, which should just be the same `_train` instead
- [x] Comments about "weights uniform" and "hidden seed"?
- [x] Why do we use `concat` to add the predictions? Can't we just create a copy of the new columns the regular way in the original dataframe (if it is OK to modify at this point)? Or use `df.assign(pred=df['predictions'])` to just print it without overwriting the variable (if we have taught `assign` before here.
    - [x] We should also not overwrite variable names like this in my opinion.
    - [x] Maybe something like this:
        ```py
        cancer_test_predictions = cancer_test.copy()
        cancer_test_predictions['predicted'] = knn_fit.predict(cancer_test.drop(columns='Class'))
        ```
- [x] X_test is created in the accuracy section but was needed in the previous code cell.
- [x] I think we can do a better job explaining accuracy and start by computed it manually using `y_test` and the predictions created in `cancer_test_predictions`, so that students get a better grasp of what it is.
- [x] Missing actual value in "The output shows that the estimated accuracy of the classifier on the test data was %."
    - [x] I just realized that all of these might be fixed when the code actually runs...
- [x] Do we ever even need to show `confusion_matrix` when we are saying that `CnfusionMatrixDisiplay` is the recommend way?
    - [x] This is one of the many places where it feels unnecessarily noise to assign to a variable instead of just displaying the plot.
- [x] We need to explain confusion matrices and what is in each square, there are resources in BAIT 509 and DSCI 573 for this.
- [x] We talk about "majority classifier" but we never show how to create one. This is really easy to setup in sklearn and it is a useful strategy for students to learn so that they can compare against a baseline. I suggest we show them how to setup a dummy classifier (and a dummy regressor in later chapters). We might also consider renaming "majority classifier" to "dummy classifier" in the text so not be ambiguous
- [x] The code in the cross-validation section repeats some of the things that were done earlier in the chapter like the train_test_split
- [x] Cross-validation fig72 legend overlaps equation
- [x] Quoting the docs for the `cv` param of `cross_validate` does not seem that helpful?
- [x] The first time we use cross_validate, we do many things at once (cros_validate, reutrn_train, convert to pandas dataframe, set cv to 5 although its the default). It could be better to do these one at a time.
- [x] "We can then aggregate the mean and standard **deviation**". Again Search and replace in the entire chapter.
    - [x] We don't need `func=` in `agg`
- [x] "provides **two** build-in methods"
- [x] I think we should introduce grid search and randomized search one at a time to make it easier to understand and also include more explanation as well as an illustration. We can take this from BAIT 509 https://bait509-ubc.github.io/BAIT509/lectures/lecture6.html#automated-hyperparameter-optimization
- [x] Do we cover what `range` does somewhere or do we just start using it?
- [x] "the **cross-validation** accuracy actually starts to decrease!
- [x] `param_grid_lots` is not a great name, something like `large_param_grid` sounds better. same with the plots name.
- [x] Do we explain `n_jobs=-1`?
- [x] Rewrite this "We can evaluate a classifier by splitting the data randomly into a training and test data set, using the training set to build the classifier, and using the test set to estimate its accuracy. Finally, we can tune the classifier (e.g., select the number of neighbors in K-NN) by maximizing estimated accuracy via cross-validation." **to this** "We can tune and evaluate a classifier by splitting the data randomly into a training and test data set. The training set is used to build the classifier and we can tune the classifier (e.g., select the number of neighbors in K-NN) by maximizing estimated accuracy via cross-validation. After we have tuned the model we can use the test set to estimate its accuracy."
- [x] Swap step 2 and 3, since 2 and 4 belong together and 3 is independent of the two.
- [x] Change step 4 to read: "Setup GridSearchCV (or RandomizedSearchCV) to estimate the classifier accuracy for a range of parameter values (values of "K" in our case)."
- [x] Change step 5 to read "Execute the grid search by passing the training data to the `fit()` function of GridSearchCV instance.
- [x] "yields a high **cross-validation accuracy** ~~estimate~~ that"
- [x] Relating to step 6 and 7, GridSearchCV already provides us with the best model as an attribute, granted hat we still want them to look at the plot, is there value in teaching how to extract the already fit best model from `GridSearchCV.best_estimator_` instead of refitting it (which is redundant and potentially time consuming).
- [x] We don't show how we create the irrelevant columns
- [x] "Forward selection ~~in Python~~"
    - [x] In general I think this section is not that well-written and we should go through it and improve the language as well as how the different code snippets are introduced.

### PR DESCRIPTION
**Pull request by trevorcampbell**
- closes #70 
- closes #9

### CODE REVIEW

**Comment by trevorcampbell** [comment]
@joelostblom there is one figure at the end (Figure 6.7) that is not generating properly --- see the R version to see what the plot should look like). Your help there would be much appreciated.

Aside from that @joelostblom and @lheagy this is ready for review.

Please review the Ch5 PR first #78 since this builds on that one.

File: source/classification2.md [hunk 1]
```diff
@@ -56,7 +42,7 @@ By the end of the chapter, readers will be able to do the following:
 - Describe what training, validation, and test data sets are and how they are used in classification.
 - Split data into training, validation, and test data sets.
 - Describe what a random seed is and its importance in reproducible data analysis.
-- Set the random seed in Python using the `numpy.random.seed` function or `random_state` argument in some of the `scikit-learn` functions.
```
Comment (joelostblom) [inline source/classification2.md:59]: See the note I added in https://github.com/UBC-DSCI/introduction-to-datascience-python/issues/114 for some more thoughts on this.

File: source/classification2.md [hunk 2]
```diff
@@ -317,23 +320,24 @@ np.random.seed(1)
...
-cancer["Class"] = cancer["Class"].apply(
-    lambda x: "Malignant" if (x == "M") else "Benign"
-)
+# re-label Class 'M' as 'Malignant', and Class 'B' as 'Benign',
+# and change the Class variable to have a category type
+cancer['Class'] = cancer['Class'].replace({
+				'M' : 'Malignant',
+				'B' : 'Benign'
+				})
+cancer['Class'] = cancer['Class'].astype('category')
```
Comment (joelostblom) [inline source/classification2.md]:
```suggestion
cancer['Class'] = cancer['Class'].replace({
    'M' : 'Malignant',
    'B' : 'Benign'
}).astype('category')
```

File: source/classification2.md [hunk 3]
```diff
@@ -180,115 +164,134 @@ The trick is that in Python&mdash;and other programming languages&mdash;randomne
...
 
 np.random.seed(1)
-random_numbers = np.random.choice(range(10), size=10, replace=True)
-random_numbers
-```
 
-You can see that `random_numbers` is a list of 10 numbers
+nums_0_to_9 = pd.Series([0, 1, 2, 3, 4, 5, 6, 7, 8, 9])
+
+random_numbers1 = nums_0_to_9.sample(n = 10).values
```
Comment (joelostblom) [inline source/classification2.md]: `values` will be deprecated in favor of `to_numpy()` so it would be preferred if we use that I think

File: source/classification2.md [hunk 4]
```diff
@@ -317,23 +320,24 @@ np.random.seed(1)
...
+cancer['Class'] = cancer['Class'].astype('category')
 
 # create scatter plot of tumor cell concavity versus smoothness,
 # labeling the points be diagnosis class
-## create a list of colors that will be used to customize the color of points
-colors = ["#86bfef", "#efb13f"]
 
 perim_concav = (
     alt.Chart(cancer)
     .mark_point(opacity=0.6, filled=True, size=40)
```
Comment (joelostblom) [inline source/classification2.md]:
```suggestion
    .mark_circle()
```

File: source/classification2.md [hunk 5]
```diff
@@ -434,42 +419,30 @@ glue("cancer_test_nrow", len(cancer_test))
...
+in `cancer_train`. We see about {glue:}`cancer_train_b_prop`% of the training
 data are benign and {glue:}`cancer_train_m_prop`% 
 are malignant, indicating that our class proportions were roughly preserved when we split the data.
 
 ```{code-cell} ipython3
-cancer_proportions = pd.DataFrame()
-cancer_proportions['n'] = cancer_train.groupby('Class')['ID'].count()
-cancer_proportions['percent'] = 100 * cancer_proportions['n'] / len(cancer_train)
-cancer_proportions
+cancer_train["Class"].value_counts(normalize = True)
```
Comment (joelostblom) [inline source/classification2.md]:
```suggestion
cancer_train["Class"].value_counts(normalize=True)
```

File: source/classification2.md [hunk 6]
```diff
@@ -546,80 +506,50 @@ knn_fit
...
-# )
-
-# # move the `predicted` column to the first column for easy visualization
-# col_order = cancer_test_predictions.columns.tolist()
-# col_order = col_order[-1:] + col_order[:-1]
-# cancer_test_predictions[col_order]
+cancer_test_predictions = cancer_test.assign(
+                  predicted = knn_fit.predict(
+                         cancer_test.loc[:, ["Smoothness", "Concavity"]]
+                  ))
```
Comment (joelostblom) [inline source/classification2.md]:
```suggestion
cancer_test_predictions = cancer_test.assign(
    predicted = knn_fit.predict(cancer_test.loc[:, ["Smoothness", "Concavity"]])
)
```

File: source/classification2.md [hunk 7]
```diff
@@ -546,80 +506,50 @@ knn_fit
...
+Finally, we can assess our classifier's accuracy. We could compute the accuracy manually
+by using our earlier formula: the number of correct predictions divided by the total
+number of predictions. First we filter the rows to find the number of correct predictions,
+and then divide the number of rows with correct predictions by the total number of rows
+using the `len` function.
 ```{code-cell} ipython3
-:tags: [remove-cell]
+correct_preds = cancer_test_predictions[
+    cancer_test_predictions['Class'] == cancer_test_predictions['predicted']
+    ]
```
Comment (joelostblom) [inline source/classification2.md]:
```suggestion
]
```

File: source/classification2.md [hunk 8]
```diff
@@ -546,80 +506,50 @@ knn_fit
...
 
-# Finally, we can assess our classifier's accuracy. To do this we use the `metrics` function \index{tidymodels!metrics}
-# from `tidymodels` to get the statistics about the quality of our model, specifying
-# the `truth` and `estimate` arguments:
+len(correct_preds) / len(cancer_test_predictions)
```
Comment (joelostblom) [inline source/classification2.md]:
```suggestion
correct_preds.shape[0] / cancer_test_predictions.shape[0]
```

File: source/classification2.md [hunk 9]
```diff
@@ -546,80 +506,50 @@ knn_fit
...
+the `score` method. To use the `score` method, we need to specify two arguments:
+predictors and true labels. We pass the same test data
+for the predictors that we originally passed into `predict` when making predictions,
+and we provide the true labels via the `cancer_test["Class"]` series.
 
+```{code-cell} ipython3
+cancer_acc_1 = knn_fit.score(
+                cancer_test.loc[:, ["Smoothness", "Concavity"]],
+                cancer_test["Class"]
+              )
```
Comment (joelostblom) [inline source/classification2.md]:
```suggestion
cancer_acc_1 = knn_fit.score(
    cancer_test.loc[:, ["Smoothness", "Concavity"]],
    cancer_test["Class"]
)
```

File: source/classification2.md [hunk 10]
```diff
@@ -629,56 +559,35 @@ cancer_acc_1
...
-```{code-cell} ipython3
-from sklearn.metrics import ConfusionMatrixDisplay
-
-confusion_display = ConfusionMatrixDisplay(
-    confusion_matrix=confusion, display_labels=knn_fit.classes_
-)
-confusion_display.plot();
+pd.crosstab(cancer_test_predictions["Class"],
+            cancer_test_predictions["predicted"]
+           )
```
Comment (joelostblom) [inline source/classification2.md]:
```suggestion
pd.crosstab(
    cancer_test_predictions["Class"],
    cancer_test_predictions["predicted"]
)
```

File: source/classification2.md [hunk 11]
```diff
@@ -732,7 +641,7 @@ As an example, in the breast cancer data, recall the proportions of benign and m
 observations in the training data are as follows:
 
 ```{code-cell} ipython3
-cancer_proportions
+cancer_train["Class"].value_counts(normalize = True)
```
Comment (joelostblom) [inline source/classification2.md]:
```suggestion
cancer_train["Class"].value_counts(normalize=True)
```

File: source/classification2.md [hunk 12]
```diff
@@ -818,92 +728,67 @@ models, and evaluate their accuracy. We will start with just a single
...
-X_valid = cancer_validation.loc[:, ["Smoothness", "Concavity"]]
-y_valid = cancer_validation["Class"]
-acc = knn_fit.score(X_valid, y_valid)
+knn_fit = make_pipeline(cancer_preprocessor, knn).fit(X, y)
 
+# compute the score on validation data
+acc = knn_fit.score(
+        cancer_validation.loc[:, ["Smoothness", "Concavity"]],
+        cancer_validation["Class"]
+       )
```
Comment (joelostblom) [inline source/classification2.md]:
```suggestion
acc = knn_fit.score(
    cancer_validation.loc[:, ["Smoothness", "Concavity"]],
    cancer_validation["Class"]
)
```

File: source/classification2.md [hunk 13]
```diff
@@ -1385,10 +1234,9 @@ completely different.  In general, if the model *is influenced too much* by the
 training data, it is said to **overfit** the data.
 
 ```{code-cell} ipython3
-:tags: [remove-cell]
 
-# create the scatter plot
-colors = ["#86bfef", "#efb13f"]
+alt.data_transformers.disable_max_rows()
+alt.renderers.enable("mimetype")
```
Comment (joelostblom) [inline source/classification2.md]:
```suggestion
```

This uses jupyterlab to insert a png as a backup image for charts and a vega-lite instead of html representation of the interactive version of the chart. Since jupyter-book does not call jupyterlab in anyway, the image is not create and the browser can't handle the vega-lite spec, but needs the html version so we'll leave it with the original renderer for now.

**Review by joelostblom — CHANGES_REQUESTED** [review]
Looking good overall! My biggest comment is the one I made in the issues regarding randomness.

File: source/classification2.md (hunk 1 above)
Comment (trevorcampbell) [reply source/classification2.md:59]: I'm going to stick with the current `np.random.seed` approach as discussed in #114

File: source/classification2.md (hunk 3 above)
Comment (trevorcampbell) [reply source/classification2.md]: good call!
